server/issue.go

```diff
@@ -774,123 +779,77 @@ func (s *Server) getPipelineCreateForDatabaseSchemaAndDataUpdate(ctx context.Con
 			}
-
-			vt, err := getUpdateTask(database, c.VCSPushEvent, d, getOrDefaultSchemaVersion(d))
+			matrix, err := getDatabaseMatrixFromDeploymentSchedule(deploySchedule, "" /* baseDatabaseName */, "" /* databaseNameTemplate */, []*api.Database{database})
 			if err != nil {
-				return nil, err
+				return nil, echo.NewHTTPError(http.StatusInternalServerError, "Failed to create deployment pipeline").SetInternal(err)
 			}
-
-			key := envKey{name: database.Instance.Environment.Name, id: database.Instance.Environment.ID, order: database.Instance.Environment.Order}
-			envToDatabaseMap[key] = append(envToDatabaseMap[key], vt)
+			for i, databaseList := range matrix {
+				aggregatedMatrix[i] = append(aggregatedMatrix[i], databaseList...)
+			}
+			databaseToMigrationList[database.ID] = append(databaseToMigrationList[database.ID], d)
 		}
-		// Sort and group by environments.
-		var envKeys []envKey
-		for k := range envToDatabaseMap {
-			envKeys = append(envKeys, k)
+	}
+
+	for i, databaseList := range aggregatedMatrix {
+		// Skip the stage if the stage includes no database.
+		if len(databaseList) == 0 {
+			continue
 		}
-		sort.Slice(envKeys, func(i, j int) bool {
-			return envKeys[i].order < envKeys[j].order
-		})
-		for _, env := range envKeys {
-			taskList, taskIndexDagList := getTaskAndDagListByVersion(envToDatabaseMap[env])
-			create.StageList = append(create.StageList, api.StageCreate{
-				Name:             env.name,
-				EnvironmentID:    env.id,
-				TaskList:         taskList,
-				TaskIndexDAGList: taskIndexDagList,
+		var environmentID int
+		var taskCreateList []api.TaskCreate
+		var taskIndexDAGList []api.TaskIndexDAG
+		for _, database := range databaseList {
+			if environmentID > 0 && environmentID != database.Instance.EnvironmentID {
+				return nil, echo.NewHTTPError(http.StatusInternalServerError, "all databases in a stage should have the same environment")
+			}
+			environmentID = database.Instance.EnvironmentID
+
+			migrationDetailList := databaseToMigrationList[database.ID]
+			sort.Slice(migrationDetailList, func(i, j int) bool {
+				return migrationDetailList[i].SchemaVersion < migrationDetailList[j].SchemaVersion
 			})
+			for i := 0; i < len(migrationDetailList)-1; i++ {
+				taskIndexDAGList = append(taskIndexDAGList, api.TaskIndexDAG{FromIndex: len(taskCreateList) + i, ToIndex: len(taskCreateList) + i + 1})
+			}
+			for _, migrationDetail := range migrationDetailList {
+				taskCreate, err := getUpdateTask(database, c.VCSPushEvent, migrationDetail, getOrDefaultSchemaVersion(migrationDetail))
+				if err != nil {
+					return nil, err
+				}
+				taskCreateList = append(taskCreateList, taskCreate)
+			}
 		}
-	}
-	return create, nil
-}
 
-// getTaskAndDagListByVersion adds the task dependencies for tasks belonging to the same database.
-// The dependency is by schema version ascending order.
-func getTaskAndDagListByVersion(versionTaskList []*versionTask) ([]api.TaskCreate, []api.TaskIndexDAG) {
```

Comment:
By removing this, we couldn't unit test the logic?

Reply:
This requires less unit testing and the version dependency is covered by TestVCS Go integration test (that's how I found a bug before).